src/main/java/org/jboss/modules/PathUtils.java

```diff
@@ -144,23 +148,52 @@ public static String fileNameOfPath(String path) {
         return path.substring(max(path.lastIndexOf('/'), path.lastIndexOf(File.separatorChar)) + 1);
     }
 
+    /**
+     * Request a new buffer at or longer than the given length. The minimum length is
+     * provided at the class level to make sure that the buffer will always exceed the minimum length.
+     *
+     * If a size requested is greater than the minimum length it will go to the next power of two up
+     * to the maximum integer value. This is intended to prevent an ascending series of paths ("/a/a.txt",
+     * "/a/aa.txt", "/a/aaa.txt") from requiring constant/increasing buffer allocations.
+     *
+     * @param length of the buffer to create
+     * @return a created character buffer that may not be empty but that is at least as long as the requeted length
+     */
+    static char[] getBuffer(final int length) {
+        char[] current = CHAR_BUFFER_CACHE.get();
+        if(current == null || current.length < length) {
+            // since we can't shift left from the max value it will hold at the max value
+            // it uses the outer Math.max to belt-and-suspenders out of a lot of corner cases that end up negative.
+            // the shifting logic finds out how many powers of two we need to shift left to get to the next power of two
+            // requesting a size that is already a power of two will give you that same power of two
+            // because we get the same power of two we use a Math.max on the inner value to make sure we always meet the minimum
+            int nexLength = length == Integer.MAX_VALUE ? Integer.MAX_VALUE : Math.max(MIN_LENGTH, 1 << (Integer.SIZE - Integer.numberOfLeadingZeros(Math.max(length, MIN_LENGTH) - 1)));
+            CHAR_BUFFER_CACHE.set(new char[nexLength]);
+        }
+        return CHAR_BUFFER_CACHE.get();
+    }
+
     /**
      * Canonicalize the given path.  Removes all {@code .} and {@code ..} segments from the path.
      *
      * @param path the relative or absolute possibly non-canonical path
      * @return the canonical path
      */
     public static String canonicalize(String path) {
+        if (path.trim().isEmpty() || (!path.contains("..") && !path.contains("."))) {
+            return path;
+        }
+        if (path.contains(".") && !path.contains("/.") && !path.contains("./")) {
+            return path;
+        }
```

Comment:
I would expect that these five whole-string scans would be more costly than just running the single-pass scan of the canonicalization routine, assuming that the changes to lazily acquire (or even just lazily allocate) the target buffer are retained. I would leave these two `if` conditions off.

Reply:
I had a JMH benchmark that I didn't want to introduce to the code base but my read of the test results was that it was quite a bit faster. I think we can also cut down the number of `contains` calls by a few and still get the same affect. 

I'll bring the benchmark back into this MR so we can compare. There's a good chance I'm reading it wrong.

Reply:
I've reorganized the results a bit but here's the summary:
```
Benchmark                                      Mode  Cnt  Score    Error  Units
PathUtilsBenchmarkTest.changeString            avgt    4  0.136 �  0.004  us/op
PathUtilsBenchmarkTest.directChangeString      avgt    4  0.136 �  0.001  us/op

PathUtilsBenchmarkTest.manyChangeString        avgt    4  0.088 �  0.002  us/op
PathUtilsBenchmarkTest.directManyChangeString  avgt    4  0.088 �  0.001  us/op

PathUtilsBenchmarkTest.noChangeString          avgt    4  0.003 �  0.001  us/op
PathUtilsBenchmarkTest.directNoChangeString    avgt    4  0.130 �  0.003  us/op

PathUtilsBenchmarkTest.withDotButFine          avgt    4  0.003 �  0.001  us/op
PathUtilsBenchmarkTest.directWithDotButFine    avgt    4  0.054 �  0.005  us/op
```

In short there isn't much difference between ones that go through the if blocks and ones that don't but the one's that get out early are much much faster.

Reply:
With a difference like this, I guess that JIT might be coming into play, or perhaps the tight loops in `contains` are somehow optimized better than the state machine loop of the main routine.

Reply:
I wonder if replacing the `switch` with `if` would help here. The cases are few enough, and the JIT cannot profile `switch` like it can with `if`.

Reply:
I think the method itself is fast enough it's just allocation pressure that slows it down. I'm also worried about the profiling and the fact that JMH does try and consume the value but the compiler might be too smart. I'm adding a few more test cases for the benchmark as well.

Reply:
These changes really affect `isChild` as well, which is the most common entry point into `canonicalize` during the boot up process (from what i am reading). I was looking at that method to see if there were any gains to be made there but I can't see any way to get around modifying the string or anything like that. I'll probably create a benchmark case for that as well but that's probably going too far.

Reply:
Ok, so this is a lot but I've added the GC profiling to JMH benchmarking and extended the benchmark duration. It takes a minute to run and can be accessed with the `benchmark` profile. I have denoted the original algorithm, a direct call without filtering ("direct"), and the pre-checking methods in the table below.

| Case | Metric | Original | Direct | Pre-Checking |
|---|---|---|---|---|
| No Change String | us/op | 0.086 | 0.093 | 0.012 |
| | allocation (MB/sec) | 8536 | 3609 | 0.001 |
| | allocation (B/op) | 192 | 88 | 10^-6 |
| Changed String | us/op | 0.104 | 0.111 | 0.122 |
| | allocation (MB/sec) | 7072 | 2740 | 2496 |
| | allocation (B/op) | 192 | 80 | 80 |
| Many Change String | us/op | 0.070 | 0.063 | 0.066 |
| | allocation (MB/sec) | 8264 | 2906 | 2770 |
| | allocation (B/op) | 152 | 48 | 48 |
| With Dot But No Change | us/op | 0.063 | 0.040 | 0.009 |
| | allocation (MB/sec) | 9185 | 6785 | 0.001 |
| | allocation (B/op) | 152 | 72 | 10^-6 |
| Long String With Change | us/op | 1.433 | 1.635 | 1.571 |
| | allocation (MB/sec) | 3278 | 1007 | 1048 |
| | allocation (B/op) | 1232 | 432 | 432 |
| Long String With No Change | us/op | 1.536 | 1.610 | 1.684 |
| | allocation (MB/sec) | 3100 | 1042 | 996 |
| | allocation (B/op) | 1248 | 440 | 440 |

If you look at the long string with a change (390 characters that doesn't have a `/.` until the end) you can see in the results that the difference between the original and pre-checking is around 14% increase. I think the JVM is outsmarting us or something though because the _direct_ benchmarks are slower than the pre-checking benchmarks sometimes. 

The allocation rate is really what makes this worthwhile anyway. A lot of these test cases have 30% of the allocations compared to the original.

The real question is if a 10-14% decrease in speed is worth it for a large speedup (~7x) when canonicalization is definitely not needed. I don't know if I'm qualified to answer that. We also have to pay for the interaction with the ThreadLocal which is around 5% in speed over the original.